___

<a href='https://www.udemy.com/user/joseportilla/'><img src='../Pierian_Data_Logo.png'/></a>
___
<center><em>Content Copyright by Pierian Data</em></center>

# **String Formatting**

String formatting lets you inject items into a string rather than trying to chain items together using commas or string concatenation. As a quick comparison, consider:

```py
player = 'Thomas'
points = 33
    
'Last night, '+player+' scored '+str(points)+' points.'  # concatenation
    
f'Last night, {player} scored {points} points.'          # string formatting
```


There are three ways to perform string formatting.
* The oldest method involves placeholders using the modulo `%` character.
* An improved technique uses the `.format()` string method.
* The newest method, introduced with Python 3.6, uses formatted string literals, called *f-strings*.

Since you will likely encounter all three versions in someone else's code, we describe each of them here.

Table of Content:
1. `String Formatting with % Operators`
2. `Formatting with the .format() method`
3. `Formatted String Literals (f-strings)`

## **Formatting with placeholders**
## **String Formatting with `%` Operators**
Old string formatting, you can use <code>%s</code> to inject strings into your print statements. The modulo `%` is referred to as a "string formatting operator".

Here are some commonly used placeholders:

1. `%s` : string
2. `%d` : integer
3. `%f` : float
4. `%.nf` : float with n digits after the decimal point

In [1]:
print("I'm going to inject %s here." %'something')

I'm going to inject something here.


You can pass multiple items by placing them inside a tuple after the `%` operator.

In [2]:
print("I'm going to inject %s text here, and %s text here." %('some','more'))

I'm going to inject some text here, and more text here.


You can also pass variable names:

In [3]:
x, y = 'some', 'more'
print("I'm going to inject %s text here, and %s text here."%(x,y))

I'm going to inject some text here, and more text here.


### **Format conversion methods.**
It should be noted that two methods <code>%s</code> and <code>%r</code> convert any python object to a string using two separate methods: `str()` and `repr()`. We will learn more about these functions later on in the course, but you should note that `%r` and `repr()` deliver the *string representation* of the object, including quotation marks and any escape characters.

In [4]:
print('He said his name was %s.' %'Fred')
print('He said his name was %r.' %'Fred')

He said his name was Fred.
He said his name was 'Fred'.


As another example, `\t` inserts a tab into a string.

In [5]:
print('I once caught a fish %s.' %'this \tbig')
print('I once caught a fish %r.' %'this \tbig')

I once caught a fish this 	big.
I once caught a fish 'this \tbig'.


The `%s` operator converts whatever it sees into a string, including integers and floats. The `%d` operator converts numbers to integers first, without rounding. Note the difference below:

In [6]:
print('I wrote %s programs today.' %3.75)
print('I wrote %d programs today.' %3.75)   

I wrote 3.75 programs today.
I wrote 3 programs today.


### **Padding and Precision of Floating Point Numbers**
Floating point numbers use the format <code>%5.2f</code>. Here, <code>5</code> would be the minimum number of characters the string should contain; these may be padded with whitespace if the entire number does not have this many digits. Next to this, <code>.2f</code> stands for how many numbers to show past the decimal point. Let's see some examples:

In [7]:
print('Floating point numbers: %5.2f' %(13.144))

Floating point numbers: 13.14


In [8]:
print('Floating point numbers: %1.0f' %(13.144))

Floating point numbers: 13


In [9]:
print('Floating point numbers: %1.5f' %(13.144))

Floating point numbers: 13.14400


In [10]:
print('Floating point numbers: %10.2f' %(13.144))

Floating point numbers:      13.14


In [11]:
print('Floating point numbers: %25.2f' %(13.144))

Floating point numbers:                     13.14


For more information on string formatting with placeholders visit https://docs.python.org/3/library/stdtypes.html#old-string-formatting

### **Multiple Formatting**
Nothing prohibits using more than one conversion tool in the same print statement:

In [12]:
print('First: %s, Second: %5.2f, Third: %r' %('hi!',3.1415,'bye!'))

First: hi!, Second:  3.14, Third: 'bye!'


## **Formatting with the `.format()` method**
A better way to format objects into your strings for print statements is with the string `.format()` method. 

The syntax is: `'String here {} then also {}'.format('something1','something2')`
    
For example:

In [13]:
print('This is a string with an {}'.format('insert'))

This is a string with an insert


### **The .format() method has several advantages over the %s placeholder method:**

#### 1. Inserted objects can be called by index position:

In [14]:
print('The {2} {1} {0}'.format('fox','brown','quick'))

The quick brown fox


#### 2. Inserted objects can be assigned keywords:

In [15]:
print('First Object: {a}, Second Object: {b}, Third Object: {c}'.format(a=1,b='Two',c=12.3))

First Object: 1, Second Object: Two, Third Object: 12.3


#### 3. Inserted objects can be reused, avoiding duplication:

In [16]:
print('A %s saved is a %s earned.' %('penny','penny'))
# vs.
print('A {p} saved is a {p} earned.'.format(p='penny'))

A penny saved is a penny earned.
A penny saved is a penny earned.


### **Alignment, padding and precision with `.format()`**
Within the curly braces you can assign field lengths, left/right alignments, rounding parameters and more

In [12]:
print('{0:8} | {1:9}'.format('Fruit', 'Quantity'))
print('{0:8} | {1:9}'.format('Apples', 3.))
print('{0:8} | {1:9}'.format('Oranges', 10))

Fruit    | Quantity 
Apples   |       3.0
Oranges  |        10


By default, `.format()` aligns text to the left, numbers to the right. You can pass an optional `<`,`^`, or `>` to set a left, center or right alignment:

In [18]:
print('{0:<8} | {1:^8} | {2:>8}'.format('Left','Center','Right'))
print('{0:<8} | {1:^8} | {2:>8}'.format(11,22,33))

Left     |  Center  |    Right
11       |    22    |       33


You can precede the aligment operator with a padding character

In [19]:
print('{0:=<8} | {1:-^8} | {2:.>8}'.format('Left','Center','Right'))
print('{0:=<8} | {1:-^8} | {2:.>8}'.format(11,22,33))

Left==== | -Center- | ...Right
11====== | ---22--- | ......33


Field widths and float precision are handled in a way similar to placeholders. The following two print statements are equivalent:

In [16]:
print('This is my ten-character, two-decimal number:%10.2f' %13.579)
print('This is my ten-character, two-decimal number:{0:10.2f}'.format(13.579))

This is my ten-character, two-decimal number:     13.58
This is my ten-character, two-decimal number:     13.58


Note that there are 5 spaces following the colon, and 5 characters taken up by 13.58, for a total of ten characters.

For more information on the string `.format()` method visit https://docs.python.org/3/library/string.html#formatstrings

## **Formatted String Literals (f-strings)**

Introduced in Python 3.6, f-strings offer several benefits over the older `.format()` string method described above. For one, you can bring outside variables immediately into to the string rather than pass them as arguments through `.format(var)`.

#### **Print Values of Variables and Evaluate Expressions**

In any f-String, `{var_name}`, `{expression}` serve as placeholders for variables and expressions, and are replaced with the corresponding values at runtime.

In [7]:
name = 'Fred'

print(f"He said his name is {name}.")

He said his name is Fred.


In [9]:
num1 = 10000
num2 = 9
print(f"The product of {num1} and {num2} is {num1 * num2}")
print(f"The product of {num1} and {num2} is {num1 * num2:,}")

The product of 10000 and 9 is 90000
The product of 10000 and 9 is 90,000


#### **Using an Object’s String Representations**

Python’s f-strings support two flags with special meaning in the interpolation process. These flags are closely related to how Python manages the string representation of objects. These flags are:

<style>
.center{
    text-align: center;
}
</style>

<table>
    <tr>
        <th>Flag</th>
        <th>Description</th>
    </tr>
    <tr>
        <td class="center">!s</td>
        <td>Interpolates the string representation from the .__str__() method, calls str()</td>
    </tr>
    <tr>
        <td class="center">!r</td>
        <td>Interpolates the string representation from the .__repr__() method, calls repr()</td>
    </tr>
    <tr>
        <td class="center">!a</td>
        <td>Calls ascii()</td>
    </tr>
</table>

In [17]:
class Person:
    def __init__(self, name, age):
        self.name = name
        self.age = age
    
    def __str__(self):
        return f"I'm {self.name}, and I'm {self.age} years old"
    
    def __repr__(self):
        return f"{type(self).__name__}(name='{self.name}', age={self.age})"

In [18]:
daffa = Person("Daffa", 22)
print(f"{daffa!s}.")

I'm Daffa, and I'm 22 years old.


In [21]:
print(f"{daffa!r}")

Person(name='Daffa', age=22)


In [32]:
print(f"He said his name is {name!r}")
print(f"He said his name is {repr(name)}.")

He said his name is 'Fred'
He said his name is 'Fred'.


### **Data Types**
There are many ways to represent strings and numbers when using f-strings. 

<style>
.center{
    text-align: center;
}
</style>

<table>
    <tr>
        <th>Type</th>
        <th>Meaning</th>
    </tr>
    <tr>
        <td class="center">s</td>
        <td>String format--this is the default type for strings</td>
    </tr>
    <tr>
        <td class="center">d</td>
        <td>Decimal integer. This uses a comma as the number separator character</td>
    </tr>
    <tr>
        <td class="center">n</td>
        <td>Number. This is the same as <b>d</b> except that it uses the current locale setting to insert the appropriate number separator characters</td>
    </tr>
    <tr>
        <td class="center">f</td>
        <td>Fixed-point notation. Displays the number as a fixed-point number. The default precision is 6</td>
    </tr>
    <tr>
        <td class="center">%</td>
        <td>Percentage. Multiplies the number by 100 and displays in fixed (<b>f</b>) format, followed by a percent sign.</td>
    </tr>
    <tr>
        <td class="center">e</td>
        <td>Exponent notation. Prints the number in scientific notation using the letter <b>e</b> to indicate the exponent. The default precision is 6.</td>
    </tr>
</table>

#### **String formatting follows**

In [33]:
variable = "10"
# variable = 10 # (int) -> ERRROR
print(f"There are {variable:s} pieces apples")

There are 10 pieces apples


### **The following is a basic example of the use of f-strings with numbers:**

#### **Integer formatting follows**

In [15]:
variable = 10
print(f"Using Numeric {variable = }")
print(f"This prints without formatting {variable}")
print(f"This prints with formatting {variable:d}")
print(f"This prints also with formatting {variable:n}")
print(f"This prints with spacing {variable:10d}")

Using Numeric variable = 10
This prints without formatting 10
This prints with formatting 10
This prints also with formatting 10
This prints with spacing         10


#### **Float formatting follows `"result: {value:{width}.{precision}}"`**

Where with the `.format()` method you might see `{value:10.4f}`, with f-strings this can become `{value:{10}.{6}}`


In [23]:
num = 23.45678
print("My 10 character, four decimal number is:{0:10.4f}".format(num))
print(f"My 10 character, four decimal number is:{num:{10}.{6}}")

My 10 character, four decimal number is:   23.4568
My 10 character, four decimal number is:   23.4568


Note that with f-strings, *precision* refers to the total number of digits, not just those following the decimal. This fits more closely with scientific notation and statistical analysis. Unfortunately, f-strings do not pad to the right of the decimal, even if precision allows it:

In [2]:
num = 23.45
print("My 10 character, four decimal number is:{0:10.4f}".format(num))
print(f"My 10 character, four decimal number is:{num:{10}.{6}}")

My 10 character, four decimal number is:   23.4500
My 10 character, four decimal number is:     23.45


If this becomes important, you can always use `.format()` method syntax inside an f-string:

In [46]:
num = 23.45
print("My 10 character, four decimal number is:{0:10.4f}".format(num))
print(f"My 10 character, four decimal number is:{num:10.4f}")

My 10 character, four decimal number is:   23.4500
My 10 character, four decimal number is:   23.4500


Display `+` (plus) and `-` (minus) signs

In [47]:
minus_num = -10
plus_num = 10.1234
print(f"minus = {minus_num:+d}")
print(f"minus = {plus_num:+.2f}")

minus = -10
minus = +10.12


#### **Percentage notation formatting follows**

In [14]:
variable = 4
print(f"Using Numeric {variable = }")
print(f"This prints without formatting {variable}")
print(f"This prints with percent formatting {variable:%}")
print(f"percent = {variable:.1%}")

Using Numeric variable = 4
This prints without formatting 4
This prints with percent formatting 400.000000%
percent = 400.0%


#### **Exponent notation formatting follows**

In [16]:
variable = 403267890
print(f"Using Numeric {variable = }")
print(f"This prints with exponential formatting {variable:e}")

Using Numeric variable = 403267890
This prints with exponential formatting 4.032679e+08


#### **Format another number (binary, octal, hexadecimal)**

In [34]:
num = 255

print(f"binary = {bin(num)}")
print(f"octal = {oct(num)}")
print(f"hexadecimal = {hex(num)}")

binary = 0b11111111
octal = 0o377
hexadecimal = 0xff


### **Alignment**

<style>
.center{
    text-align:center;
}
</style>

<table>
    <tr>
        <th>Option</th>
        <th>Meaning</th>
    </tr>
    <tr>
        <td class="center"><</td>
        <td>Forces the expression within the curly braces to be left-aligned. This is the default for strings.</td>
    </tr>
    <tr>
        <td class="center">></td>
        <td>Forces the expression within the curly braces to be right-aligned. This is the default for numbers.</td>
    </tr>
    <tr>
        <td class="center">^</td>
        <td>Forces the expression within the curly braces to be centered. </td>
    </tr>
</table>

The following is an example using alignment for both a number and a string. The "|" is used in
the f-string to help delineate the spacing. That number after the ":"will cause that field to be
that number of characters wide. The first line in the print() statement using f-strings is an
example of using f-strings for debugging purposes which will be covered later.

In [20]:
import math
variable = math.pi

print(f"Using Numeric {variable = }")
print(f"|{variable:25}|")
print(f"|{variable:<25}|")
print(f"|{variable:>25}|")
print(f"|{variable:^25}|\n")

variable = "Python 3.9"
print(f"Using String {variable = }")
print(f"|{variable:25}|")
print(f"|{variable:<25}|")
print(f"|{variable:>25}|")
print(f"|{variable:^25}|")

Using Numeric variable = 3.141592653589793
|        3.141592653589793|
|3.141592653589793        |
|        3.141592653589793|
|    3.141592653589793    |

Using String variable = 'Python 3.9'
|Python 3.9               |
|Python 3.9               |
|               Python 3.9|
|       Python 3.9        |


A fill character can also be used in the alignment of f-strings. This is shown in the following
example:

In [22]:
import math
variable = math.pi

print(f"Using String {variable = }")
print(f"|{variable:=<25}|")
print(f"|{variable:=>25}|")
print(f"|{variable:=^25}|\n")

variable = "Python 3.9"
print(f"Using String {variable = }")
print(f"|{variable:=<25}|")
print(f"|{variable:=>25}|")
print(f"|{variable:=^25}|")

Using String variable = 3.141592653589793
|3.141592653589793========|
|========3.141592653589793|
|====3.141592653589793====|

Using String variable = 'Python 3.9'
|Python 3.9===============|
|===============Python 3.9|
|=======Python 3.9========|


#### **Self-Documenting Expressions for Debugging**

Updated in version 3.9 of Python is the use of f-strings for debugging purposes.

F-strings have another cool feature that can be useful, especially during your debugging process. The feature helps you self-document some of your expressions. For example, say that you’re dealing with a minor bug or issue in your code, and you want to know the value of a variable at a given moment in the code’s execution.

For this quick check, you can insert a call `f'{expr=}'` to `print()` like the following:

In [31]:
variable = "Some mysterious value"

print(f"{variable =}")

variable ='Some mysterious value'


In [23]:
import math
goldenRatio = (1+math.sqrt(5))/2
print(f"{goldenRatio=}")
print(f"{goldenRatio = }")
print(f"{goldenRatio = :.6f}")
print(f"{(1+math.sqrt(5))/2 = :.6f}")

goldenRatio=1.618033988749895
goldenRatio = 1.618033988749895
goldenRatio = 1.618034
(1+math.sqrt(5))/2 = 1.618034


You can use a variable name followed by an equal sign (=) in an f-string to create a self-documented expression. When Python runs the f-string, it builds an expression-like string containing the variable’s name, the equal sign, and the variable’s current value. This f-string feature is useful for inserting quick debugging checks in your code.

**Conclusion**

Printing output in Python is facilitated with f-strings as it can be considered What You See is
What You Get (WYSIWYG). The procedure is as follows:
* Placing between the quotation marks after the 'f' the text that you want displayed
* Enclosing the variables to be displayed within the text in curly braces
* Within those curly braces, placing a colon (:) after the variable
* Formatting the variable using a format specification (width, alignment, data type) after
the colon

For more info on formatted string literals visit https://docs.python.org/3/reference/lexical_analysis.html#f-strings

That is the basics of string formatting!